File: src/main/java/com/baulsupp/oksocial/network/GoogleDns.java
package com.baulsupp.oksocial.network;

import com.baulsupp.oksocial.authenticator.AuthUtil;
import com.google.common.net.InetAddresses;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import okhttp3.Dns;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;

public class GoogleDns implements Dns {
  private List<InetAddress> dnsHosts;
  private IPvMode mode;
  private Supplier<OkHttpClient> client;

  public GoogleDns(List<InetAddress> dnsHosts, IPvMode mode, Supplier<OkHttpClient> client) {
    this.dnsHosts = dnsHosts;
    this.mode = mode;
    this.client = client;
  }

  // TODO implement DnsMode internally
  @Override public List<InetAddress> lookup(String host) throws UnknownHostException {
    if (host.equals("dns.google.com")) {
      return dnsHosts;
    }

    try {
      // TODO map punycode here?
      HttpUrl url = HttpUrl.parse("https://dns.google.com/resolve?name=" + host + "&type=" + type(mode));
      Request request = new Request.Builder().url(url).header("Accept", "application/dns+json").build();
      Map<String, Object> result = AuthUtil.makeJsonMapRequest(client.get(), request);

      return responseToList(result);
    } catch (IOException e) {
      UnknownHostException unknownHostException =
          new UnknownHostException("failed to lookup " + host + " via dns.google.com");
      unknownHostException.initCause(e);
      throw unknownHostException;
    }
  }

  private String type(IPvMode mode) {
    // TODO support IPv6 preferred etc, e.g. two queries
    switch (mode) {
      case IPV6_ONLY:
        return "AAAA";
      default:
        return "A";
    }
  }

  private List<InetAddress> responseToList(Map<String, Object> result) throws UnknownHostException {
    if (!result.get("Status").equals(0)) {
      // TODO response codes
      throw new UnknownHostException("Status from dns.google.com: " + result.get("Status"));
    }

    List<Map<String, Object>> answer = (List<Map<String, Object>>) result.get("Answer");

    return answer.stream()
        .filter(a -> a.get("type").equals(1) || a.get("type").equals(28))
        .map(a -> InetAddresses.forString((String) a.get("data")))
        .collect(
            Collectors.toList());
  }

  public static GoogleDns fromResourceList() {
    throw new UnsupportedOperationException();
  }

  public static GoogleDns fromHosts(Supplier<OkHttpClient> clientSupplier, IPvMode mode, String... ips) {
    List<InetAddress> hosts = new ArrayList<>();

    for (String ip : ips) {
      hosts.add(InetAddresses.forString(ip));
    }

    return new GoogleDns(hosts, mode, clientSupplier);
  }
}

File: src/main/java/com/baulsupp/oksocial/network/GoogleDns.kt
package com.baulsupp.oksocial.network

import com.baulsupp.oksocial.authenticator.AuthUtil
import com.google.common.net.InetAddresses
import okhttp3.Dns
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.net.InetAddress
import java.net.UnknownHostException
import java.util.function.Supplier
import kotlin.streams.toList

class GoogleDns(private val dnsHosts: List<InetAddress>, private val mode: IPvMode,
    private val client: Supplier<OkHttpClient>) : Dns {

  // TODO implement DnsMode internally
  @Throws(UnknownHostException::class)
  override fun lookup(host: String): List<InetAddress> {
    if (host == "dns.google.com") {
      return dnsHosts
    }

    try {
      // TODO map punycode here?
      val url = HttpUrl.parse("https://dns.google.com/resolve?name=" + host + "&type=" + type(mode))
      val request = Request.Builder().url(url!!).header("Accept", "application/dns+json").build()
      val result = AuthUtil.makeJsonMapRequest(client.get(), request)

      return responseToList(result)
    } catch (e: IOException) {
      val unknownHostException = UnknownHostException("failed to lookup $host via dns.google.com")
      unknownHostException.initCause(e)
      throw unknownHostException
    }

  }

  private fun type(mode: IPvMode): String {
    // TODO support IPv6 preferred etc, e.g. two queries
    when (mode) {
      IPvMode.IPV6_ONLY -> return "AAAA"
      else -> return "A"
    }
  }

  @Throws(UnknownHostException::class)
  private fun responseToList(result: Map<String, Any>): List<InetAddress> {
    if (result["Status"] != 0) {
      // TODO response codes
      throw UnknownHostException("Status from dns.google.com: " + result["Status"])
    }

    val answer = result["Answer"] as List<Map<String, Any>>

    return answer.stream()
        .filter { a -> a["type"] == 1 || a["type"] == 28 }
        .map { a -> InetAddresses.forString(a["data"] as String) }
        .toList()
  }

  companion object {

    fun fromResourceList(): GoogleDns {
      throw UnsupportedOperationException()
    }

    fun fromHosts(clientSupplier: Supplier<OkHttpClient>, mode: IPvMode,
        vararg ips: String): GoogleDns {
      val hosts = ips.map { InetAddresses.forString(it) }

      return GoogleDns(hosts, mode, clientSupplier)
    }
  }
}
